### DISCUSSION TITLE
Request for Bulk Refresh Capability for SQLAlchemy Objects

### DISCUSSION THREAD

**Discussion by AliHezarpisheh** [sqlalchemy/sqlalchemy#11488] [Ideas]
## Problem Statement:
When utilizing methods like bulk_save_objects for bulk insertion or modification in SQLAlchemy, managing the refresh of these objects afterward is cumbersome and requires manual iteration through the list of objects. This process could be streamlined with a dedicated method for bulk refresh operations.

## Proposed Solution:
Introduce a new method, bulk_refresh(instances: list[objects]), that efficiently handles the refresh of a list of SQLAlchemy objects after bulk operations. This method would simplify the current practice of individually iterating over each object for refresh post bulk operations.

Introducing a bulk_refresh method would provide a better developer experience for the alchemists!

**Comment by zzzeek** [comment]
hi -

bulk_save_objects is a legacy method first off, but also I dont think the proposal makes sense.  the objects you pass to bulk_save_objects are not part of the session, therefore there is no "refresh".   The whole point of "bulk insert" is performance, so you would not want to be emitting SELECT for all those objects afterwards.  If you have a bunch of objects that you want to INSERT and then have them in your transaction with all your data and so-forth, just use the normal unit of work patterns. (I've noticed people using bulk_save_objects because they were not even aware of `session.add()` ).

if you want to bulk insert objects and get back ORM-attached objects with everything as it is in the database, use [ORM bulk INSERT with RETURNING](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#getting-new-objects-with-returning).     this allows you to write an efficient INSERT that will get the data inserted and selected at the same time.

**Reply by AliHezarpisheh** [reply]
Hello,

Thank you for getting back to me and for the clarifications provided.

I understand now that bulk_save_objects is more of a legacy method and not intended for scenarios where post-insertion refreshes are required, given that it doesn't include the objects in the session.

The suggestion to use ORM bulk INSERT with RETURNING is very helpful. This approach seems to align better with my goal of efficiently inserting many objects and having them readily available within the session afterward.
